messaging/src/main/java/io/lantern/messaging/CryptoWorker.kt

```diff
@@ -0,0 +1,837 @@
+                        }
+
+                        val finalAttachmentBuilder = if (isThumbnail)
+                            readAttachment.toBuilder().setThumbnail(attachmentBuilder.build())
+                        else
+                            attachmentBuilder
+                        msgBuilder.putAttachments(id, finalAttachmentBuilder.build())
+                    }
+
+                    val updatedMsg = msgBuilder.build()
+                    tx.put(msgPath, updatedMsg)
+
+                    if (updatedMsg.allAttachmentsUploaded) {
+                        submit { processOutbound(out) }
+                    }
+                }
+            }
+        }
+
+        val ciphertextLength =
+            AttachmentCipherOutputStream.getCiphertextLength(attachment.attachment.plaintextLength)
+        if (ciphertextLength > auth.maxUploadSize) {
+            // TODO: cleanly handle case when attachment exceeds allowed size, including proper notification to user
+            logger.error("attachment size exceeds allowed size of ${auth.maxUploadSize}, failing")
+            updateStatus(false)
+            return
+        }
+
+        val requestBody = MultipartBody.Builder().setType(MultipartBody.FORM)
+        auth.uploadFormDataMap.forEach { (key, value) ->
+            requestBody.addFormDataPart(key, value)
+        }
+        requestBody.addFormDataPart(
+            "file",
+            "filename",
+            File(attachment.encryptedFilePath).asRequestBody()
+        )
+        val rb = requestBody.build()
+        val request = Request.Builder().url(auth.uploadURL).post(rb).build()
+        httpClient.newCall(request).enqueue(object : okhttp3.Callback {
+            override fun onResponse(call: Call, response: Response) {
+                submit {
+                    try {
+                        val success = response.code == 204
+                        if (!success) {
+                            logger.error("upload failed with unretriable status ${response.code}: ${response.body?.string()}") // ktlint-disable max-line-length
+                        }
+                        updateStatus(success)
+                    } finally {
+                        response.body?.close()
+                    }
+                }
+            }
+
+            override fun onFailure(call: Call, e: IOException) {
+                logger.error("failed to upload attachment, will try again: ${e.message}")
+                retryFailed { uploadAttachment(out, msg, id, attachment, isThumbnail) }
+            }
+        })
+    }
+
+    internal fun getMoreUploadAuthorizationsIfNecessary(then: () -> Unit = {}) {
+        removeExpiredUploadAuthorizations()
+        val numToRequest = 10 - uploadAuthorizations.size
+        if (numToRequest < 0) {
+            then()
+            return
+        }
+
+        messaging.anonymousClientWorker.withClient { client ->
+            client.requestUploadAuthorizations(
```

Comment:
[This test](https://github.com/getlantern/tassis/blob/main/attachments/s3attachments/s3attachments_test.go) may explain it better than me, but ...

Uploads are stored in an S3-compatible service (Wasabi). They're stored in a bucket that's world readable. However, we don't want to make it world writeable because then anybody could store anything they want in there, of any size, not to mention deleting or modifying existing data.

So, we use a [pre-signed post](https://bigbinary.com/blog/uploading-files-directly-to-s3-using-pre-signed-post-request) whereby we can limit the authorization simply to uploading new content, limited to 100MB.

It does occur to me that we should add logic to ensure we're not overwriting existing content. Logged https://github.com/getlantern/tassis/issues/9.

Reply:
Ahh okay.  Interesting!  Thanks for the explanation.